lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -422,6 +422,121 @@ static bool sinkCode(Function *F) {
+/// Check the use-def dependency up to a depth of \p depth.
+static bool mayDepend(Node *A, Node *B, unsigned depth = 6) {
+  // We define the identify as a dependency.
+  if (A == B) {
+    return true;
+  }
+
+  // A does not depend on anything.
+  if (A->getNumInputs() == 0) {
+    return false;
+  }
+
+  // B has no users. Nothing can depend on it.
+  if (B->getNumResults() == 0) {
+    return false;
+  }
+
+  // We can't continue the search. Assume that the nodes depend on one another.
+  if (depth == 0) {
+    return true;
+  }
+
+  // Check all inputs of A. None of them may depend on B.
+  for (int i = 0, e = A->getNumInputs(); i < e; i++) {
+    auto *input = A->getNthInput(i).getNode();
+    // The inputs of A must not depend on B.
+    if (mayDepend(input, B, depth - 1)) {
+      return true;
+    }
+  }
+
+  // We checked all inputs of A and none of them depend on B.
+  return false;
+}
+
+/// \returns True if the node \p N depends on any of the values in \p list, or
+/// if any of the values in list depend on \p N.
+static bool mayDependOnAny(llvm::ArrayRef<NodeValue> list, Node *N) {
+  for (auto ll : list) {
+    if (mayDepend(ll.getNode(), N) || mayDepend(N, ll.getNode())) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+// Merge several two or more multiple matrix multiplications into a single
+// large matmul. The large matmul is more likely to utilize the hardware. The
+// result of the big matmul is the concatenated results.
+//
+//            ____      _________        _________
+//   ----    |    |    |         |     M|  A * C  |
+// M| A  |  T| B  | * K|    C    | =    |---------|
+//   ---- ,  |    |    |         |     T|  B * C  |
+//    K       ----      ---------        ---------
+//             K            R                R
+static void mergeMatMul(Function *F) {
+  auto &nodes = F->getNodes();
+
+  // These two maps record the list of matrix multipliers that use each node
+  // value either as a right-hand-side user or a left-hand-user.
+  llvm::DenseMap<Node *, std::vector<MatMulNode *>> righMatrixUsers;
+  llvm::DenseMap<Node *, std::vector<MatMulNode *>> leftMatrixUsers;
+
+  // Collect the list of nodes that are used by the matrix multiplier.
+  for (auto const &node : nodes) {
+    if (auto *MM = dyn_cast<MatMulNode>(node)) {
+      righMatrixUsers[MM->getRHS().getNode()].push_back(MM);
+      leftMatrixUsers[MM->getLHS().getNode()].push_back(MM);
+    }
+  }
+
+  // Merge RHS matrices.
+  for (auto &it : righMatrixUsers) {
+    auto &MMs = it.second;
+
+    // Collects the LHS values to merge.
+    std::vector<NodeValue> LHS;
+
+    // For each matmul that depends on the rhs matrix.
+    for (auto &MM : MMs) {
+      auto L = MM->getLHS();
+      // The operands to the matrix multiplier should not depend on one another
+      // or else we won't be able to get rid of the original matrix
+      // multiplication.
+      if (mayDependOnAny(LHS, L.getNode())) {
```

Comment:
If the default depth of 6 for `mayDepend()` is too small and a dependency does exist, would an assert somewhere later on catch this?

Reply:
@jfix71 It's okay for mayDepend to fail. This is a compile-time vs performance tradeoff. If we can't trace the graph to the depth of 6 then we should probably not try to analyze this region and give up. I think that in practice this won't be a real performance problem. Heuristics like this are used in LLVM and GCC.

Reply:
@jfix71  Looks like depth == 6 is a good constant to use  :)

https://github.com/llvm-mirror/llvm/blob/c36d516fd6b1c105a53b3dd22843979e85ad6383/lib/Transforms/InstCombine/InstCombineSimplifyDemanded.cpp#L127

Reply:
Ah yeah it conservatively returns true if it gets to max depth.